### PR TITLE
removes custom scalers from kubectl

### PR DESCRIPTION
**Pull request by p0lyn0mial**
**What this PR does / why we need it**: this PR removes custom scalers from kubectl and uses the genericScaler instead.

**Release note**:

```
NONE
```

### CODE REVIEW

File: pkg/kubectl/testing/fake.go
Comment (p0lyn0mial) [inline pkg/kubectl/testing/fake.go]: yeah, I can hear you :) How bad this actually is ?

FYI: https://github.com/kubernetes/kubernetes/issues/56865

File: pkg/kubectl/cmd/util/factory_client_access.go
Comment (deads2k) [inline pkg/kubectl/cmd/util/factory_client_access.go]: move this out to the ring2Factory and it can be no-arg.  See the example here: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/factory_builder.go#L46

File: pkg/kubectl/cmd/util/factory_object_mapping.go
Comment (deads2k) [inline pkg/kubectl/cmd/util/factory_object_mapping.go]: The `Scaler` function can temporarily move to `ring2Factory`.  It should have very few callers.  Open an issue to remove all of them.

File: pkg/kubectl/cmd/util/factory_object_mapping.go
Comment (deads2k) [inline pkg/kubectl/cmd/util/factory_object_mapping.go]: @pwittrock @adohe @soltysh I think we need to start trying to kill reapers.  GC doesn't resolve problems for objects that are nested (remove this SA subject from that rolebinding), but we should start thinking of alternatives.

File: pkg/kubectl/cmd/util/factory_object_mapping.go
Comment (deads2k) [inline pkg/kubectl/cmd/util/factory_object_mapping.go]: Move out to ring 2.

File: pkg/kubectl/delete.go
Comment (deads2k) [inline pkg/kubectl/delete.go]: Create a `NewScaler(scalesGetter scaleclient.ScalesGetter, gr schema.GroupResource)` function.  When scaling jobs is dead, we'll remove `ScalerFor`

Comment (deads2k) [reply pkg/kubectl/delete.go]: > When scaling jobs is dead, we'll remove ScalerFor

And when that dies, we'll update the `Scaler` interface to accept `GroupResource` as an arg so we don't have to choose ahead of time.

File: pkg/kubectl/testing/fake.go
Comment (deads2k) [reply pkg/kubectl/testing/fake.go]: > yeah, I can hear you :) How bad this actually is ?

@DirectXMan12 We really need a fake scale client we can use.  Is it just a matter of finding time?

File: staging/src/k8s.io/client-go/scale/client.go
Comment (deads2k) [inline staging/src/k8s.io/client-go/scale/client.go]: I'm guessing you changed this to stop masking errors so that `IsNotFound` and the like would work?  If so, please add a comment.

**Comment by deads2k** [comment]
As much as I want it, I don't think we're going to be landing this for 1.10.

/cc @kubernetes/sig-cli-maintainers

File: pkg/kubectl/testing/fake.go
Comment (DirectXMan12) [reply pkg/kubectl/testing/fake.go]: Yeah, IIRC, I have one that was usuable with a fake rest mapper, but it's a bit clunky.  Let me try and take some time today and make it better.

Comment (DirectXMan12) [reply pkg/kubectl/testing/fake.go]: https://github.com/p0lyn0mial/kubernetes/pull/2
^---------- the current fake scale client works fine for this

Comment (DirectXMan12) [reply pkg/kubectl/testing/fake.go]: that should also close #56865

Comment (p0lyn0mial) [reply pkg/kubectl/testing/fake.go]: @DirectXMan12 thanks for showing me how to use the fake client :) I think it will work and we can remove this custom code.

**Comment by p0lyn0mial** [comment]
@deads2k @DirectXMan12 I think I have addressed all the comments. PTAL

File: pkg/kubectl/cmd/util/factory.go
Comment (deads2k) [inline pkg/kubectl/cmd/util/factory.go]: This level of the factory can use the lower factory levels to get its mapping

Comment (deads2k) [reply pkg/kubectl/cmd/util/factory.go]: > This level of the factory can use the lower factory levels to get its mapping

Nm, wrong mapping mapper thing

File: pkg/kubectl/scale.go
Comment (deads2k) [inline pkg/kubectl/scale.go]: @p0lyn0mial are you up for removing this in a follow-up pull.  @soltysh got it deprecated in 1.10, so we'll be good to remove in 1.11.

**Comment by deads2k** [comment]
Great pull. @soltysh will be really happy about the tests.

1.10 is in freeze, so I've set the milestone for 1.11.

/lgtm

File: pkg/kubectl/scale.go
Comment (p0lyn0mial) [reply pkg/kubectl/scale.go]: sure, no problem, does it mean that we actually want to have a scale endpoint for jobs ?

https://github.com/kubernetes/kubernetes/pull/58468

Comment (deads2k) [reply pkg/kubectl/scale.go]: > sure, no problem, does it mean that we actually want to have a scale endpoint for jobs ?
> (#58468)

No.  We'll remove the special case here and you'll no longer be able to scale them from kubectl since it wasn't scaling in the same sense.

File: pkg/kubectl/cmd/util/factory_builder.go
Comment (soltysh) [inline pkg/kubectl/cmd/util/factory_builder.go]: Move this check right after `f.clientAccessFactory.ClientSet()`. What's the point of continuing if we failed getting clientset. Generally fail fast is the desired approach.

Comment (soltysh) [reply pkg/kubectl/cmd/util/factory_builder.go]: I see it's copy&paste, but sine you've touched it, can you please fix it :wink:

**Review by soltysh — APPROVED** [review]
This lgtm, but I'd like to have that one tiny bit fixed before merging it.
/lgtm cancel

File: pkg/kubectl/cmd/testing/BUILD
Comment (mengqiy) [inline pkg/kubectl/cmd/testing/BUILD]: Please revert this change.
I guess this is happening because you delete and regenerate it?

File: pkg/kubectl/cmd/testing/BUILD
Comment (mengqiy) [inline pkg/kubectl/cmd/testing/BUILD]: Same here.

File: pkg/kubectl/cmd/util/BUILD
Comment (mengqiy) [inline pkg/kubectl/cmd/util/BUILD]: Same here.

File: pkg/kubectl/cmd/util/BUILD
Comment (mengqiy) [inline pkg/kubectl/cmd/util/BUILD]: Same here.

File: pkg/kubectl/delete.go
Comment (mengqiy) [inline pkg/kubectl/delete.go]: nit: revert this.

File: pkg/kubectl/cmd/testing/BUILD
Comment (p0lyn0mial) [reply pkg/kubectl/cmd/testing/BUILD]: yeah, usually when bazel complains and regenerating doesn't help as a last resort I delete `BUILD` files 😇

File: pkg/kubectl/cmd/util/factory_builder.go
Comment (p0lyn0mial) [reply pkg/kubectl/cmd/util/factory_builder.go]: sure, no problem.

**Comment by p0lyn0mial** [comment]
@soltysh @mengqiy comments addressed.

**Review by soltysh — APPROVED** [review]
/lgtm

**Comment by soltysh** [comment]
@p0lyn0mial mind rebasing this one/

**Comment by janetkuo** [comment]
test-cmd starts flaking around the same time this PR got merged: #61748

Would you take a look to see if it's related?

**Comment by deads2k** [comment]
> test-cmd starts flaking around the same time this PR got merged: #61748
> 
> @p0lyn0mial Would you take a look to see if it's related?

It looks like k8s.io/kubernetes/test/integration/deployment.TestDeploymentLabelAdopted started failing at the same time. That doesn't use the scaler right?

**Pull request merged**
